### ISSUE TITLE
eid table failed to sync up

### ISSUE THREAD

**Issue by drakedog2008** [CodeConstruct/mctp#1]
mtcpd cannot allocate correct eid when setting up new end point with multiple controllers. 

It looks like `n->peeridx[eid]` does sync up with dbus obj tree. 

```
busctl call xyz.openbmc_project.MCTP  /xyz/openbmc_project/mctp  au.com.CodeConstruct.MCTP  SetupEndpoint say mctpi2c8 1 0x1d
Call failed: Endpoint claimed EID 9 which is already used

mctp addr
eid 8 net 1 dev mctpi2c3
eid 9 net 1 dev mctpi2c5
eid 10 net 1 dev mctpi2c6
eid 11 net 1 dev mctpi2c7
eid 12 net 1 dev mctpi2c8
eid 13 net 1 dev mctpi2c9
eid 14 net 1 dev mctpi2c10
eid 15 net 1 dev mctpi2c11

mctp link
dev lo index 1 address 0x00:00:00:00:00:00 net 1 mtu 65536 up
dev mctpi2c3 index 4 address 0x10 net 1 mtu 254 up
dev mctpi2c5 index 5 address 0x10 net 1 mtu 254 up
dev mctpi2c6 index 6 address 0x10 net 1 mtu 254 up
dev mctpi2c7 index 7 address 0x10 net 1 mtu 254 up
dev mctpi2c8 index 8 address 0x10 net 1 mtu 254 up
dev mctpi2c9 index 9 address 0x10 net 1 mtu 254 up
dev mctpi2c10 index 10 address 0x10 net 1 mtu 254 up
dev mctpi2c11 index 11 address 0x10 net 1 mtu 254 up

busctl tree xyz.openbmc_project.MCTP
`-/xyz
  `-/xyz/openbmc_project
    `-/xyz/openbmc_project/mctp
      `-/xyz/openbmc_project/mctp/1
        |-/xyz/openbmc_project/mctp/1/10
        |-/xyz/openbmc_project/mctp/1/11
        |-/xyz/openbmc_project/mctp/1/12
        |-/xyz/openbmc_project/mctp/1/13
        |-/xyz/openbmc_project/mctp/1/14
        |-/xyz/openbmc_project/mctp/1/15
        |-/xyz/openbmc_project/mctp/1/8
        `-/xyz/openbmc_project/mctp/1/9
```

**Comment by drakedog2008** [comment]
https://github.com/CodeConstruct/mctp/blob/c858b0fed919a4c35410a67b87764c3422b5bcaa/src/mctpd.c#L1196-L1204

It looks like `peeridx[9]` remains -1 in the search loop.

**Comment by drakedog2008** [comment]
Hi @jk-ozlabs 

Thanks for the reply. 

>the device you're enumerating (at 0x1d on mctpi2c8) already has an EID of 9
net 1 eid 9 is not taken by end point enumeration, but it is manually allocated to a controller of another i2c bus(mcpti2c5). 

```
eid 9 net 1 dev mctpi2c5
eid 12 net 1 dev mctpi2c8
```

We have a script on boot to allocate eid/s to controllers of each PE i2c. In this case, eid 8-15 are taken, so end point enumeration should start with eid 16?

**Comment by jk-ozlabs** [comment]
Hi @drakedog2008 

> net 1 eid 9 is not taken by end point enumeration, but it is manually allocated to a controller of another i2c bus(mcpti2c5).

That's correct - you're not assigning that EID yourself, but the drive already had that EID - possibly left-over from an earlier BMC boot where you only had one interface (likely mctpi2c8) defined.


> ```
> eid 9 net 1 dev mctpi2c5
> eid 12 net 1 dev mctpi2c8
> ```
> 
> We have a script on boot to allocate eid/s to controllers of each PE i2c. In this case, eid 8-15 are taken, so end point enumeration should start with eid 16?

Yes, that's what will happen; the EID pool will start from the lowest *available* address, so EIDs 8-15 would not be assigned. We just need to ensure that the drive is using EIDs from that pool - and using `AssignEndpoint` rather than `SetupEndpoint` would achieve that.

**Comment by drakedog2008** [comment]
> possibly left-over from an earlier BMC boot where you only had one interface (likely mctpi2c8) defined.

Why reboot will have left-over on the EID assignment? I thought mtcpd will rescan/update the EID table at start up? 

> and using AssignEndpoint rather than SetupEndpoint would achieve that.

I will try `AssignEndpoint`. But wouldn't `SetupEndpoint` be the superset of `AssignEndpoint`? In another word,  `LearnEndpoint` returns false since `mctp addr` didn't show eid allocation for the endpiont device.

**Comment by jk-ozlabs** [comment]
@mkj has made some changes to make `SetupEndpoint` more robust for external device state - these are included v1.0, so closing this. Let me know how that goes.
